Refactor app access

GavinFrazar opened the pull request:
This PR refactors app access to move the logic in the AWS handler and Azure handler out of `RoundTrip` and into `ServeHTTP`, since we were not following the RoundTripper interface requirements in multiple ways. Also, `oxy.Forwarder` documents that middleware should be done using ServeHTTP not in RoundTrip.

This PR also moves the AWS signing service into `lib/utils/aws/signing.go` so that it can be re-used for database-access DynamoDB in a future PR.

I tested that app access including AWS and Azure still work.

GavinFrazar committed:
Refactor app access

* Move logic out of RoundTrip and into ServeHTTP as a middleware before handing off to oxy forwarder
* Move AWS signing service code into lib/utils/aws/signing.go

GavinFrazar commented:
@Tener requesting your review since this PR affects the Azure integration you just added. As mentioned above I tested that Azure integration still works with an Azure VM, managed identity, and `tsh az ...`.

GavinFrazar committed:
Merge branch 'master' into gavinfrazar/refactor-app-access-middleware

jakule commented on lib/srv/app/common/audit.go:
```diff
@@ -177,14 +177,14 @@ func (a *audit) OnRequest(ctx context.Context, sessionCtx *SessionContext, req *
 		Path:               req.URL.Path,
 		RawQuery:           req.URL.RawQuery,
-		StatusCode:         uint32(res.StatusCode),
+		StatusCode:         uint32(code),
```
Instead of casting here, can we change the function argument? I think it makes more sense to accept the argument of expected type and cast when a function is called when needed.

GavinFrazar replied:
Good suggestion. I wonder why golang's `ResponseWriter` takes an int status code anyway? Shouldn't status > 0 always?

jakule commented on lib/srv/app/common/audit.go:
```diff
@@ -203,7 +203,7 @@ func (a *audit) OnDynamoDBRequest(ctx context.Context, sessionCtx *SessionContex
 		AWSRequestMetadata: *MakeAWSRequestMetadata(req, re),
 		SessionChunkID:     sessionCtx.ChunkID,
-		StatusCode:         uint32(res.StatusCode),
+		StatusCode:         uint32(statusCode),
```
The same as above. I'd change the function argument type to `uint32`

jakule reviewed (approved):
PR looks fine. Accepting with a few comments.

tigrato commented on lib/srv/app/aws/handler.go:
```diff
@@ -17,211 +17,171 @@ limitations under the License.
+func rewriteRequest(r *http.Request, re *endpoints.ResolvedEndpoint) (*http.Request, error) {
+	// shallow copy request and make a deep copy for header modification.
+	outReq := &http.Request{}
+	*outReq = *r
```
can't you just use `r.Clone()`?
I am afraid that some other fields are missing from this clone and can cause bugs in the future.

It also clones the `r.URL` so you can safely manipulate it in `urlForResolvedEndpoint`

GavinFrazar replied:
good catch, it "works" but definitely looks bug-prone.

How's this instead:
```golang
// rewriteRequest rewrites a request to remove Teleport reserved headers, sets the url, and sets host.
func rewriteRequest(ctx context.Context, r *http.Request, re *endpoints.ResolvedEndpoint) (*http.Request, error) {
	u, err := urlForResolvedEndpoint(r, re)
	if err != nil {
		return nil, trace.Wrap(err)
	}
	outReq, err := http.NewRequestWithContext(ctx, r.Method, u.String(), r.Body)
	if err != nil {
		return nil, trace.Wrap(err)
	}
	outReq.Header = r.Header.Clone()
...
```

GavinFrazar replied:
`ctx` in this case would be the app server close context by the way.

GavinFrazar replied:
See above comments. I will change this to use clone instead, as you suggested. This stuff is subtle 😑

GavinFrazar replied:
updated to clone and assign the clone's body to a new readcloser.

tigrato commented on lib/utils/aws/signing.go:
```diff
@@ -0,0 +1,194 @@
+	}
+
+	// copy removed headers back to the request after signing it, but don't copy the old Authorization header.
+	copyHeaders(reqCopy, req, utils.RemoveFromSlice(unsignedHeaders, "Authorization"))
```
+1

GavinFrazar replied:
I think that suggestion won't work, because we do not want all of the original request headers when we sign. If we sign with those headers, then they become signed headers as well.

Maybe I've misunderstood the suggestion though?

smallinsky replied:
Yes, that true, but why we need to add a extra headers to reqCopy form the origin req after signing reqCopy ? 
The previous version of this code the reqCopy  was jus returned:  
https://github.com/gravitational/teleport/pull/19387/files#diff-e247bd4fbfee0e25ceae30d0afae77f5f3e68f661470d5ead8519042c333ca7eL204

GavinFrazar replied:
In this PR, the signing service is decoupled from the app service AWS handler. Previously, we were removing app service `ReservedHeaders` and then not adding them back after signing. That is still the case, except now app service headers are removed by the app service AWS handler before giving the request to the signing service for signature. So that behavior is actually the same before/after: we remove app service headers before signing and do not copy them back.

What I've done here in the signing service is I've avoided any signature validation problems caused by signing an AWS request with headers that should not be signed, such as hop-by-hop headers (which may change before reaching AWS for signature validation). It's just a robust way of signing forwarded AWS requests. Previously, the oxy forwarder would remove hop-by-hop headers for us before we got the request to sign, but that is no longer the case since we are not signing inside `RoundTrip`.

I tested the difference in wire format sent before/after this PR just now, and there is a difference in the headers we send, which is unrelated to the signing service changes here. Before this PR, we had AWS handling logic inside `RoundTrip`, which was being called by `oxy/forward/Forwarder.ServeHTTP`. The oxy forwarder rewrites headers before calling `RoundTrip` though, using this default rewriter in `oxy/forward/rewrite.go`:
```golang
// Rewriter is responsible for removing hop-by-hop headers and setting forwarding headers
type HeaderRewriter struct {
	TrustForwardHeader bool
	Hostname           string
}
```
This means that before this PR, those forwarding headers were already injected into the request by oxy and were then removed by the AWS signing roundtripper. I think that is a mistake - we recently changed `ReservedHeaders` to include the headers injected by oxy, thus we started to remove those oxy forwarding headers. This PR fixes that - since this PR moves the AWS handling logic out of `RoundTrip`, those forwarding headers are injected by oxy after we sign the request, and so we are now sending the forwarding headers again.

Here's an example AWS request in wire format before and after these changes so you can see which headers are being sent now vs before.

Before this PR:
```
POST / HTTP/1.1
Host: localhost:8001
Accept-Encoding: identity
Authorization: AWS4-HMAC-SHA256 Credential=*snip*, SignedHeaders=accept-encoding;content-type;host;x-amz-date;x-amz-security-token;x-amz-target, Signature=*snip*
Content-Length: 2
Content-Type: application/x-amz-json-1.0
User-Agent: aws-cli/2.9.7 Python/3.11.0 Darwin/22.2.0 source/arm64 prompt/off command/dynamodb.list-tables
X-Amz-Date: 20221219T190825Z
X-Amz-Security-Token: *snip*
X-Amz-Target: DynamoDB_20120810.ListTables

{}
```

This PR:
```
POST / HTTP/1.1
Host: localhost:8001
Transfer-Encoding: chunked
Accept-Encoding: identity
Authorization: AWS4-HMAC-SHA256 Credential=*snip*, SignedHeaders=content-type;host;x-amz-date;x-amz-security-token;x-amz-target, Signature=*snip*
Content-Type: application/x-amz-json-1.0
User-Agent: aws-cli/2.9.7 Python/3.11.0 Darwin/22.2.0 source/arm64 prompt/off command/dynamodb.list-tables
X-Amz-Date: 20221219T185744Z
X-Amz-Security-Token: *snip*
X-Amz-Target: DynamoDB_20120810.ListTables
X-Forwarded-Host: localhost:8001
X-Forwarded-Port: 8001
X-Forwarded-Proto: http
X-Forwarded-Server: mac.attlocal.net

2
{}
0
```

A few things to note here about the request sent using this PR's changes:
1. `X-Forwarded-*` headers injected by oxy are now present in the request, which I believe is correct behavior because a reverse proxy should be setting those headers.
2. Content-Length is not set, so go automatically set the `Transfer-Encoding: chunked` header for us. The original request was not chunked, so this is a mistake I made. That value was actually coming from `http.Request.ContentLength`, not `http.Request.Header.Get("Content-Length")`. In fact the problem is the way I copy the request by making a new request and passing the request body directly. I will fix this by cloning the request in the AWS handler instead.
3. The signed headers no longer includes `accept-encoding`, which it shouldn't since this was not a signed header originally.

tl;dr: We remove unsigned headers and copy them back after signing so that we can sign the request without unintentionally signing other headers. I need to fix the request copying in `rewriteRequest` to use `req.Clone` instead of making a new request so that we preserve the content length in the request.

GavinFrazar replied:
Correction: we were not sending `x-forwarded-*` headers before the changes to `ReservedHeaders` either... I assumed we only removed those headers for signing purposes. Is there any reason we shouldn't allow the oxy forwarder to set those headers?

smallinsky replied:
IFIK no, these headers should be removed only for signing purpose.

tigrato replied:
I agree with @smallinsky
We should include them after we sign the request

GavinFrazar replied:
Latest changes continue to send the oxy-injected forwarding headers and fixed the transfer encoding issue, so the request is not chunked now.

tigrato commented on lib/utils/aws/signing.go:
```diff
@@ -0,0 +1,194 @@
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	reqCopy, err := http.NewRequest(req.Method, req.URL.String(), bytes.NewReader(payload))
```
`req.Clone()` followed by `req.Body = ...`

GavinFrazar replied:
I don't think the `req.Body` change is needed since we called `GetAndReplaceReqBody(req)` already?

GavinFrazar replied:
nvm I think it's safer to just assign to body after cloning

GavinFrazar commented on lib/srv/app/aws/handler_test.go:
```diff
@@ -213,7 +215,7 @@ func TestAWSSignerHandler(t *testing.T) {
 			}
 
-			suite := createSuite(t, handler, tc.app)
+			suite := createSuite(t, mockAWSHandler, tc.app, clockwork.NewRealClock())
```
fake clock works, we just don't use it anywhere in the test. I'll change to a fake clock since it works though

GavinFrazar commented on lib/srv/app/aws/handler.go:
```diff
@@ -17,211 +17,171 @@ limitations under the License.
+		auditErr = sessCtx.Audit.OnRequest(unsignedReq.Context(), sessCtx, unsignedReq, recorder.Status(), re)
 	}
-	reqCopy, err := http.NewRequest(r.Method, url, bytes.NewReader(payload))
 	if err != nil {
```
good catch!

GavinFrazar replied:
I've also added `closeContext` to the handler since a cancelled request context should not affect audit log event emitting. It gets the close context from the app server, just like we do in the Azure handler.

GavinFrazar committed:
address feedback

* check auditErr instead of err for logging
* use app server close context for audit event emitting
* add go doc comments.
* refactor request rewriting to make the copy in a more robust way.
* pass status code as uint32 rather than casting in audit emitter
* clone request in signing service

GavinFrazar committed:
clone request instead of making a new request, and rewrite url to force https

GavinFrazar committed:
Update header handling

The handlers for aws/azure were inside of an oxy/forward.Forwarder
RoundTrip function but once moved outside of that we should not pass
host header of the inbound request.

* Set oxy forwarder to PassHostHeader=false to ensure the host header is
  the URL being sought.
* Remove code that deleted forwarding headers previously, we should keep
  those (X-Forwarded-*).
* Audit log the AWS Host sought rather than the incoming request Host
  header (prior behavior maintained, we just rewrite the request
  differently using Clone).

GavinFrazar commented:
want to give @Tener a change to take a look when he's back before I merge this

GavinFrazar committed:
Merge branch 'master' into gavinfrazar/refactor-app-access-middleware

Tener reviewed (approved):
Thank you for this refactor, much appreciated.

GavinFrazar committed:
Remove obsolete header copying helper func

GavinFrazar committed:
Merge branch 'master' into gavinfrazar/refactor-app-access-middleware

GavinFrazar committed:
Merge branch 'master' into gavinfrazar/refactor-app-access-middleware

GavinFrazar committed:
Merge branch 'master' into gavinfrazar/refactor-app-access-middleware
